collaborative-plonk/src/round5.rs

```diff
@@ -0,0 +1,429 @@
+use crate::{
+    plonk_utils,
+    round3::FinalPolys,
+    round4::{Round4Challenges, Round4Proof},
+    types::Keccak256Transcript,
+    Domains, FieldShare, FieldShareVec, PlonkData, PlonkProofResult,
+};
+use ark_ec::pairing::Pairing;
+use ark_ff::Field;
+use circom_types::{
+    plonk::PlonkProof,
+    traits::{CircomArkworksPairingBridge, CircomArkworksPrimeFieldBridge},
+};
+use mpc_core::traits::{
+    FFTPostProcessing, FFTProvider, MSMProvider, PairingEcMpcProtocol, PrimeFieldMpcProtocol,
+};
+use num_traits::One;
+use num_traits::Zero;
+
+pub(super) struct Round5<T, P: Pairing>
+where
+    for<'a> T: PrimeFieldMpcProtocol<P::ScalarField>
+        + PairingEcMpcProtocol<P>
+        + FFTProvider<P::ScalarField>
+        + MSMProvider<P::G1>
+        + MSMProvider<P::G2>,
+    P::ScalarField: mpc_core::traits::FFTPostProcessing,
+{
+    pub(super) driver: T,
+    pub(super) domains: Domains<P>,
+    pub(super) challenges: Round4Challenges<P>,
+    pub(super) proof: Round4Proof<P>,
+    pub(super) polys: FinalPolys<T, P>,
+    pub(super) data: PlonkData<T, P>,
+}
+pub(super) struct Round5Challenges<P: Pairing> {
+    beta: P::ScalarField,
+    gamma: P::ScalarField,
+    alpha: P::ScalarField,
+    xi: P::ScalarField,
+    v: [P::ScalarField; 5],
+}
+
+impl<P: Pairing> Round5Challenges<P> {
+    fn new(round4_challenges: Round4Challenges<P>, v: [P::ScalarField; 5]) -> Self {
+        Self {
+            beta: round4_challenges.beta,
+            gamma: round4_challenges.gamma,
+            alpha: round4_challenges.alpha,
+            xi: round4_challenges.xi,
+            v,
+        }
+    }
+}
+
+impl<P: Pairing + CircomArkworksPairingBridge> Round4Proof<P>
+where
+    P::BaseField: CircomArkworksPrimeFieldBridge,
+    P::ScalarField: CircomArkworksPrimeFieldBridge,
+{
+    fn into_final_proof(self, commit_wxi: P::G1, commit_wxiw: P::G1) -> PlonkProof<P> {
+        PlonkProof {
+            a: self.commit_a.into(),
```

Comment:
In general, I don't like `.into()` and woud prefer `NewType::from` most of the time, even if it is more wordy, but I am open for opinions on this

Reply:
@0xThemis ?

Reply:
Rewrote it to `NewType::from`